File: tasktracker-fe/src/lib/services/userCalendarService.ts
/**
 * Copyright (c) 2025 Carlos Abril Jr
 * All rights reserved.
 *
 * This source code is licensed under the Business Source License 1.1
 * found in the LICENSE file in the root directory of this source tree.
 */

import { apiService } from './apiService';

// DTOs matching backend models
export interface UserGlobalCalendarDTO {
    userId: number;
    userName: string;
    families: UserFamilyCalendarSummaryDTO[];
    allEvents: FamilyCalendarEventWithFamilyDTO[];
    statistics: UserCalendarStatisticsDTO;
    generatedAt: string;
}

export interface FamilyCalendarEventWithFamilyDTO extends FamilyCalendarEventDTO {
    family: FamilyCalendarSummaryDTO;
    familyColor: string;
    isUserCreator: boolean;
    isUserAttendee: boolean;
    userRole: string;
}

export interface FamilyCalendarEventDTO {
    id: number;
    familyId: number;
    title: string;
    description: string;
    location: string;
    eventType: string;
    startTime: string;
    endTime: string;
    isAllDay: boolean;
    isRecurring: boolean;
    recurrencePattern?: string;
    priority: string;
    isPrivate: boolean;
    createdById: number;
    createdByName: string;
    createdAt: string;
    lastModified: string;
    attendees: FamilyCalendarEventAttendeeDTO[];
    reminders: FamilyCalendarReminderDTO[];
}

export interface FamilyCalendarEventAttendeeDTO {
    id: number;
    eventId: number;
    familyMemberId: number;
    memberName: string;
    response: string;
    responseDate?: string;
    notes?: string;
}

export interface FamilyCalendarReminderDTO {
    id: number;
    eventId: number;
    reminderTime: string;
    reminderType: string;
    message: string;
    isActive: boolean;
}

export interface UserFamilyCalendarSummaryDTO {
    familyId: number;
    familyName: string;
    familyDescription: string;
    userRole: string;
    familyColor: string;
    totalEvents: number;
    upcomingEvents: number;
    todayEvents: number;
    userCreatedEvents: number;
    userAttendingEvents: number;
    nextEventDate?: string;
    lastActivity: string;
    hasPermissionToCreateEvents: boolean;
    hasPermissionToManageEvents: boolean;
}

export interface FamilyCalendarSummaryDTO {
    id: number;
    name: string;
    description: string;
    color: string;
    memberCount: number;
}

export interface UserAvailabilityDTO {
    userId: number;
    date: string;
    isAvailable: boolean;
    busySlots: TimeSlotDTO[];
    availableSlots: TimeSlotDTO[];
    conflictingEvents: FamilyCalendarEventWithFamilyDTO[];
    timeZone: string;
}

export interface TimeSlotDTO {
    startTime: string;
    endTime: string;
    description: string;
    type: string;
    familyId?: number;
    eventId?: number;
}

export interface CalendarConflictDTO {
    id: number;
    conflictDate: string;
    conflictStartTime: string;
    conflictEndTime: string;
    conflictingEvents: FamilyCalendarEventWithFamilyDTO[];
    conflictType: string;
    severity: string;
    description: string;
    isResolved: boolean;
    detectedAt: string;
}

export interface UserCalendarStatisticsDTO {
    userId: number;
    totalFamilies: number;
    totalEvents: number;
    upcomingEvents: number;
    pastEvents: number;
    todayEvents: number;
    thisWeekEvents: number;
    thisMonthEvents: number;
    eventsCreatedByUser: number;
    eventsUserAttending: number;
    activeConflicts: number;
    resolvedConflicts: number;
    familyStats: FamilyStatisticsSummaryDTO[];
    eventTypeDistribution: { [key: string]: number };
    dailyActivityPattern: { [key: string]: number };
    hourlyActivityPattern: { [key: number]: number };
    monthlyActivityPattern: { [key: string]: number };
    averageEventsPerFamily: number;
    averageEventsPerWeek: number;
    eventAttendanceRate: number;
    eventCreationRate: number;
    lastEventCreated?: string;
    lastEventAttended?: string;
    nextUpcomingEvent?: string;
    busiestDayOfWeek: number;
    busiestHourOfDay: number;
    busiestFamily: string;
    mostCommonEventType: string;
    generatedAt: string;
}

export interface FamilyStatisticsSummaryDTO {
    familyId: number;
    familyName: string;
    familyColor: string;
    totalEvents: number;
    upcomingEvents: number;
    userCreatedEvents: number;
    userAttendingEvents: number;
    participationRate: number;
    lastActivity?: string;
    mostActiveEventType: string;
}

export interface CreateEventRequest {
    familyId: number;
    title: string;
    description: string;
    location: string;
    eventType: string;
    startTime: string;
    endTime: string;
    isAllDay: boolean;
    isRecurring: boolean;
    recurrencePattern?: string;
    priority: string;
    isPrivate: boolean;
    attendeeIds: number[];
    reminders: CreateReminderRequest[];
}

export interface CreateReminderRequest {
    reminderTime: string;
    reminderType: string;
    message: string;
}

export interface UpdateEventRequest extends CreateEventRequest {
    id: number;
}

class UserCalendarService {
    private readonly baseUrl = '/user/calendar';

    // Global calendar aggregation
    async getUserGlobalCalendar(): Promise<UserGlobalCalendarDTO | null> {
        try {
            const response = await apiService.get<UserGlobalCalendarDTO>(`${this.baseUrl}/all-families`);
            return response.data || null;
        } catch (error) {
            console.error('Error getting global calendar:', error);
            return null;
        }
    }

    async getAllUserEvents(): Promise<FamilyCalendarEventWithFamilyDTO[]> {
        try {
            const response = await apiService.get<FamilyCalendarEventWithFamilyDTO[]>(`${this.baseUrl}/all-families/events`);
            return response.data || [];
        } catch (error) {
            console.error('Error getting all user events:', error);
            return [];
        }
    }

    async getAllUserEventsInRange(startDate: string, endDate: string): Promise<FamilyCalendarEventWithFamilyDTO[]> {
        try {
            const response = await apiService.get<FamilyCalendarEventWithFamilyDTO[]>(`${this.baseUrl}/all-families/events/range`, {
                params: { startDate, endDate }
            });
            return response.data || [];
        } catch (error) {
            console.error('Error getting events in range:', error);
            return [];
        }
    }

    async getAllUserEventsToday(): Promise<FamilyCalendarEventWithFamilyDTO[]> {
        try {
            const response = await apiService.get<FamilyCalendarEventWithFamilyDTO[]>(`${this.baseUrl}/all-families/events/today`);
            return response.data || [];
        } catch (error) {
            console.error('Error getting today\'s events:', error);
            return [];
        }
    }

    async getAllUserUpcomingEvents(days: number = 7): Promise<FamilyCalendarEventWithFamilyDTO[]> {
        try {
            const response = await apiService.get<FamilyCalendarEventWithFamilyDTO[]>(`${this.baseUrl}/all-families/events/upcoming`, {
                params: { days }
            });
            return response.data || [];
        } catch (error) {
            console.error('Error getting upcoming events:', error);
            return [];
        }
    }

    // Family summaries
    async getUserFamiliesCalendarSummary(): Promise<UserFamilyCalendarSummaryDTO[]> {
        try {
            const response = await apiService.get<UserFamilyCalendarSummaryDTO[]>(`${this.baseUrl}/families-summary`);
            return response.data || [];
        } catch (error) {
            console.error('Error getting family calendar summaries:', error);
            return [];
        }
    }

    // User availability
    async getUserAvailability(date?: string): Promise<UserAvailabilityDTO | null> {
        try {
            const params = date ? { date } : {};
            const response = await apiService.get<UserAvailabilityDTO>(`${this.baseUrl}/availability`, { params });
            return response.data || null;
        } catch (error) {
            console.error('Error getting user availability:', error);
            return null;
        }
    }

    // Conflict detection
    async getUserCalendarConflicts(): Promise<CalendarConflictDTO[]> {
        try {
            const response = await apiService.get<CalendarConflictDTO[]>(`${this.baseUrl}/conflicts`);
            return response.data || [];
        } catch (error) {
            console.error('Error getting calendar conflicts:', error);
            return [];
        }
    }

    // Statistics and analytics
    async getUserCalendarStatistics(): Promise<UserCalendarStatisticsDTO | null> {
        try {
            const response = await apiService.get<UserCalendarStatisticsDTO>(`${this.baseUrl}/statistics`);
            return response.data || null;
        } catch (error) {
            console.error('Error getting calendar statistics:', error);
            return null;
        }
    }

    // Event management (delegated to family calendar service for specific families)
    async createEvent(request: CreateEventRequest): Promise<FamilyCalendarEventDTO | null> {
        try {
            const response = await apiService.post<FamilyCalendarEventDTO>(`/family/${request.familyId}/calendar/events`, request);
            return response.data || null;
        } catch (error) {
            console.error('Error creating event:', error);
            throw error;
        }
    }

    async updateEvent(request: UpdateEventRequest): Promise<FamilyCalendarEventDTO | null> {
        try {
            const response = await apiService.put<FamilyCalendarEventDTO>(`/family/${request.familyId}/calendar/events/${request.id}`, request);
            return response.data || null;
        } catch (error) {
            console.error('Error updating event:', error);
            throw error;
        }
    }

    async deleteEvent(familyId: number, eventId: number): Promise<boolean> {
        try {
            await apiService.delete(`/family/${familyId}/calendar/events/${eventId}`);
            return true;
        } catch (error) {
            console.error('Error deleting event:', error);
            return false;
        }
    }

    async updateEventAttendance(familyId: number, eventId: number, response: string, notes?: string): Promise<boolean> {
        try {
            await apiService.put(`/family/${familyId}/calendar/events/${eventId}/attendance`, {
                response,
                notes
            });
            return true;
        } catch (error) {
            console.error('Error updating event attendance:', error);
            return false;
        }
    }

    // Utility methods for frontend
    formatEventForCalendar(event: FamilyCalendarEventWithFamilyDTO) {
        return {
            id: event.id,
            title: event.title,
            start: new Date(event.startTime),
            end: new Date(event.endTime),
            allDay: event.isAllDay,
            resource: {
                familyId: event.familyId,
                familyName: event.family.name,
                familyColor: event.familyColor,
                eventType: event.eventType,
                priority: event.priority,
                isUserCreator: event.isUserCreator,
                isUserAttendee: event.isUserAttendee,
                location: event.location,
                description: event.description,
                attendees: event.attendees,
                reminders: event.reminders
            }
        };
    }

    getEventTypeColor(eventType: string): string {
        const colors: { [key: string]: string } = {
            'Meeting': '#3b82f6',
            'Appointment': '#10b981',
            'Task': '#f59e0b',
            'Reminder': '#ef4444',
            'Event': '#8b5cf6',
            'Birthday': '#ec4899',
            'Holiday': '#06b6d4',
            'Vacation': '#14b8a6',
            'Work': '#6b7280',
            'Personal': '#84cc16',
            'Family': '#f97316',
            'Social': '#8b5cf6'
        };
        return colors[eventType] || '#6b7280';
    }

    getPriorityColor(priority: string): string {
        const colors: { [key: string]: string } = {
            'Low': '#6b7280',
            'Normal': '#3b82f6',
            'High': '#f59e0b',
            'Critical': '#ef4444'
        };
        return colors[priority] || '#3b82f6';
    }

    formatConflictSeverity(severity: string): { color: string; label: string } {
        const severityMap: { [key: string]: { color: string; label: string } } = {
            'Low': { color: '#84cc16', label: 'Minor Conflict' },
            'Medium': { color: '#f59e0b', label: 'Schedule Conflict' },
            'High': { color: '#ef4444', label: 'Major Conflict' },
            'Critical': { color: '#dc2626', label: 'Critical Conflict' }
        };
        return severityMap[severity] || { color: '#6b7280', label: 'Unknown' };
    }

    isEventHappening(event: FamilyCalendarEventWithFamilyDTO): boolean {
        const now = new Date();
        const start = new Date(event.startTime);
        const end = new Date(event.endTime);
        return now >= start && now <= end;
    }

    isEventUpcoming(event: FamilyCalendarEventWithFamilyDTO, hours: number = 24): boolean {
        const now = new Date();
        const start = new Date(event.startTime);
        const hoursFromNow = new Date(now.getTime() + (hours * 60 * 60 * 1000));
        return start > now && start <= hoursFromNow;
    }

    getEventDuration(event: FamilyCalendarEventWithFamilyDTO): string {
        const start = new Date(event.startTime);
        const end = new Date(event.endTime);
        const durationMs = end.getTime() - start.getTime();
        const hours = Math.floor(durationMs / (1000 * 60 * 60));
        const minutes = Math.floor((durationMs % (1000 * 60 * 60)) / (1000 * 60));
        
        if (hours === 0) {
            return `${minutes}m`;
        } else if (minutes === 0) {
            return `${hours}h`;
        } else {
            return `${hours}h ${minutes}m`;
        }
    }
}

export const userCalendarService = new UserCalendarService(); 

File: TaskTrackerAPI/Services/UserCalendarService.cs
/*
 * Copyright (c) 2025 Carlos Abril Jr
 * All rights reserved.
 *
 * This source code is licensed under the Business Source License 1.1
 * found in the LICENSE file in the root directory of this source tree.
 *
 * This file may not be used, copied, modified, or distributed except in
 * accordance with the terms contained in the LICENSE file.
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using AutoMapper;
using Microsoft.Extensions.Logging;
using TaskTrackerAPI.DTOs.User;
using TaskTrackerAPI.DTOs.Family;
using TaskTrackerAPI.Models;
using TaskTrackerAPI.Repositories.Interfaces;
using TaskTrackerAPI.Services.Interfaces;

namespace TaskTrackerAPI.Services;

public class UserCalendarService : IUserCalendarService
{
    private readonly IUserCalendarRepository _userCalendarRepository;
    private readonly IMapper _mapper;
    private readonly ILogger<UserCalendarService> _logger;

    // Family colors for visual distinction
    private readonly string[] FamilyColors = {
        "#3B82F6", "#EF4444", "#10B981", "#F59E0B", "#8B5CF6", 
        "#06B6D4", "#F97316", "#84CC16", "#EC4899", "#6366F1"
    };

    public UserCalendarService(
        IUserCalendarRepository userCalendarRepository,
        IMapper mapper,
        ILogger<UserCalendarService> logger)
    {
        _userCalendarRepository = userCalendarRepository;
        _mapper = mapper;
        _logger = logger;
    }

    public async Task<UserGlobalCalendarDTO> GetUserGlobalCalendarAsync(int userId)
    {
        try
        {
            // Get user's aggregated calendar data
            var allEvents = await _userCalendarRepository.GetAllUserEventsAsync(userId);
            var families = await _userCalendarRepository.GetUserFamiliesAsync(userId);
            var conflicts = await _userCalendarRepository.GetUserConflictingEventsAsync(userId, 
                DateTime.Today, DateTime.Today.AddDays(30));

            // Map events with family information
            var eventsWithFamily = allEvents.Select(e => 
            {
                var eventDto = _mapper.Map<FamilyCalendarEventWithFamilyDTO>(e);
                eventDto.FamilyColor = GetFamilyColor(e.FamilyId);
                return eventDto;
            }).ToList();

            // Get statistics
            var statistics = await GetUserCalendarStatisticsAsync(userId);

            // Create global calendar DTO
            var globalCalendar = new UserGlobalCalendarDTO
            {
                UserId = userId,
                UserName = families.FirstOrDefault()?.Members?.FirstOrDefault(m => m.UserId == userId)?.User?.Username ?? "",
                Families = (await GetUserFamiliesCalendarSummaryAsync(userId)).ToList(),
                AllEvents = eventsWithFamily,
                Statistics = statistics,
                GeneratedAt = DateTime.UtcNow
            };

            return globalCalendar;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting user global calendar for user {UserId}", userId);
            throw;
        }
    }

    public async Task<IEnumerable<FamilyCalendarEventWithFamilyDTO>> GetAllUserEventsAsync(int userId)
    {
        try
        {
            var events = await _userCalendarRepository.GetAllUserEventsAsync(userId);
            
            return events.Select(e => 
            {
                var eventDto = _mapper.Map<FamilyCalendarEventWithFamilyDTO>(e);
                eventDto.FamilyColor = GetFamilyColor(e.FamilyId);
                return eventDto;
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting all user events for user {UserId}", userId);
            throw;
        }
    }

    public async Task<IEnumerable<FamilyCalendarEventWithFamilyDTO>> GetAllUserEventsInRangeAsync(int userId, DateTime startDate, DateTime endDate)
    {
        try
        {
            var events = await _userCalendarRepository.GetUserEventsInRangeAsync(userId, startDate, endDate);
            
            return events.Select(e => 
            {
                var eventDto = _mapper.Map<FamilyCalendarEventWithFamilyDTO>(e);
                eventDto.FamilyColor = GetFamilyColor(e.FamilyId);
                return eventDto;
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting user events in range for user {UserId}", userId);
            throw;
        }
    }

    public async Task<IEnumerable<FamilyCalendarEventWithFamilyDTO>> GetAllUserEventsTodayAsync(int userId)
    {
        try
        {
            var events = await _userCalendarRepository.GetUserEventsTodayAsync(userId);
            
            return events.Select(e => 
            {
                var eventDto = _mapper.Map<FamilyCalendarEventWithFamilyDTO>(e);
                eventDto.FamilyColor = GetFamilyColor(e.FamilyId);
                return eventDto;
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting user events today for user {UserId}", userId);
            throw;
        }
    }

    public async Task<IEnumerable<FamilyCalendarEventWithFamilyDTO>> GetAllUserUpcomingEventsAsync(int userId, int days = 7)
    {
        try
        {
            var events = await _userCalendarRepository.GetUserUpcomingEventsAsync(userId, days);
            
            return events.Select(e => 
            {
                var eventDto = _mapper.Map<FamilyCalendarEventWithFamilyDTO>(e);
                eventDto.FamilyColor = GetFamilyColor(e.FamilyId);
                return eventDto;
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting user upcoming events for user {UserId}", userId);
            throw;
        }
    }

    public async Task<IEnumerable<UserFamilyCalendarSummaryDTO>> GetUserFamiliesCalendarSummaryAsync(int userId)
    {
        try
        {
            var families = await _userCalendarRepository.GetUserFamiliesAsync(userId);
            var familyEventCounts = await _userCalendarRepository.GetFamilyEventCountsAsync(userId);
            var familyUpcomingCounts = await _userCalendarRepository.GetFamilyUpcomingEventCountsAsync(userId);
            var familyTodayCounts = await _userCalendarRepository.GetFamilyTodayEventCountsAsync(userId);
            var userCreatedCounts = await _userCalendarRepository.GetUserCreatedEventCountsAsync(userId);
            var userAttendingCounts = await _userCalendarRepository.GetUserAttendingEventCountsAsync(userId);
            var nextEventDates = await _userCalendarRepository.GetFamilyNextEventDatesAsync(userId);
            var lastActivityDates = await _userCalendarRepository.GetFamilyLastActivityDatesAsync(userId);

            var summaries = new List<UserFamilyCalendarSummaryDTO>();

            foreach (var family in families)
            {
                var userMember = family.Members?.FirstOrDefault(m => m.UserId == userId);
                
                var summary = _mapper.Map<UserFamilyCalendarSummaryDTO>(family);
                summary.FamilyColor = GetFamilyColor(family.Id);
                summary.TotalEvents = familyEventCounts.GetValueOrDefault(family.Id, 0);
                summary.UpcomingEvents = familyUpcomingCounts.GetValueOrDefault(family.Id, 0);
                summary.TodayEvents = familyTodayCounts.GetValueOrDefault(family.Id, 0);
                summary.UserCreatedEvents = userCreatedCounts.GetValueOrDefault(family.Id, 0);
                summary.UserAttendingEvents = userAttendingCounts.GetValueOrDefault(family.Id, 0);
                summary.NextEventDate = nextEventDates.GetValueOrDefault(family.Id);
                summary.LastActivity = lastActivityDates.GetValueOrDefault(family.Id, DateTime.MinValue);
                summary.HasPermissionToCreateEvents = await _userCalendarRepository.CanUserCreateEventsInFamilyAsync(userId, family.Id);
                summary.HasPermissionToManageEvents = await _userCalendarRepository.CanUserManageEventsInFamilyAsync(userId, family.Id);
                summary.UserRole = userMember?.Role?.Name ?? "Member";

                summaries.Add(summary);
            }

            return summaries;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting user families calendar summary for user {UserId}", userId);
            throw;
        }
    }

    public async Task<UserAvailabilityDTO> GetUserAvailabilityAsync(int userId, DateTime date)
    {
        try
        {
            var eventsOnDate = await _userCalendarRepository.GetUserEventsOnDateAsync(userId, date);
            var conflictingEvents = await _userCalendarRepository.GetUserConflictingEventsAsync(userId, 
                date.Date, date.Date.AddDays(1).AddTicks(-1));

            // Calculate busy periods
            var busyPeriods = eventsOnDate.Select(e => new TimeSlotDTO
            {
                StartTime = e.StartTime,
                EndTime = e.EndTime,
                Description = e.Title,
                Type = "Busy"
            }).ToList();

            // Calculate free periods (simplified - between 9 AM and 6 PM)
            var dayStart = date.Date.AddHours(9);
            var dayEnd = date.Date.AddHours(18);
            var freePeriods = CalculateFreePeriods(dayStart, dayEnd, busyPeriods);

            return new UserAvailabilityDTO
            {
                UserId = userId,
                Date = date,
                IsAvailable = !eventsOnDate.Any(),
                ConflictingEvents = conflictingEvents.Select(e => _mapper.Map<FamilyCalendarEventWithFamilyDTO>(e)).ToList(),
                BusySlots = busyPeriods,
                AvailableSlots = freePeriods,
                TimeZone = "UTC"
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting user availability for user {UserId} on {Date}", userId, date);
            throw;
        }
    }

    public async Task<IEnumerable<CalendarConflictDTO>> GetUserCalendarConflictsAsync(int userId)
    {
        try
        {
            var startDate = DateTime.Today;
            var endDate = startDate.AddDays(30);
            var conflictingEvents = await _userCalendarRepository.GetUserConflictingEventsAsync(userId, startDate, endDate);

            var conflicts = new List<CalendarConflictDTO>();

            // Group overlapping events into conflicts
            var eventsList = conflictingEvents.ToList();
            for (int i = 0; i < eventsList.Count; i++)
            {
                for (int j = i + 1; j < eventsList.Count; j++)
                {
                    if (DoEventsOverlap(eventsList[i], eventsList[j]))
                    {
                        var conflict = new CalendarConflictDTO
                        {
                            Id = conflicts.Count + 1,
                            ConflictDate = eventsList[i].StartTime.Date,
                            ConflictStartTime = eventsList[i].StartTime > eventsList[j].StartTime ? eventsList[i].StartTime : eventsList[j].StartTime,
                            ConflictEndTime = eventsList[i].EndTime < eventsList[j].EndTime ? eventsList[i].EndTime : eventsList[j].EndTime,
                            ConflictingEvents = new List<FamilyCalendarEventWithFamilyDTO> 
                            { 
                                _mapper.Map<FamilyCalendarEventWithFamilyDTO>(eventsList[i]),
                                _mapper.Map<FamilyCalendarEventWithFamilyDTO>(eventsList[j])
                            },
                            ConflictType = "TimeOverlap",
                            Severity = "Medium",
                            Description = $"Overlapping events: '{eventsList[i].Title}' and '{eventsList[j].Title}'",
                            IsResolved = false,
                            DetectedAt = DateTime.UtcNow
                        };

                        conflicts.Add(conflict);
                    }
                }
            }

            return conflicts.DistinctBy(c => c.Id);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting user calendar conflicts for user {UserId}", userId);
            throw;
        }
    }

    public async Task<UserCalendarStatisticsDTO> GetUserCalendarStatisticsAsync(int userId)
    {
        try
        {
            var totalEvents = await _userCalendarRepository.GetUserTotalEventsCountAsync(userId);
            var upcomingEvents = await _userCalendarRepository.GetUserUpcomingEventsCountAsync(userId);
            var pastEvents = await _userCalendarRepository.GetUserPastEventsCountAsync(userId);
            var eventsCreated = await _userCalendarRepository.GetUserEventsCreatedCountAsync(userId);
            var eventsAttending = await _userCalendarRepository.GetUserEventsAttendingCountAsync(userId);
            var eventTypes = await _userCalendarRepository.GetUserEventTypeDistributionAsync(userId);
            var dailyPattern = await _userCalendarRepository.GetUserDailyActivityPatternAsync(userId);
            var hourlyPattern = await _userCalendarRepository.GetUserHourlyActivityPatternAsync(userId);
            var monthlyPattern = await _userCalendarRepository.GetUserMonthlyActivityPatternAsync(userId);
            var lastEventCreated = await _userCalendarRepository.GetUserLastEventCreatedAsync(userId);
            var lastEventAttended = await _userCalendarRepository.GetUserLastEventAttendedAsync(userId);
            var nextUpcomingEvent = await _userCalendarRepository.GetUserNextUpcomingEventAsync(userId);

            var userFamilies = await _userCalendarRepository.GetUserFamiliesAsync(userId);

            return new UserCalendarStatisticsDTO
            {
                UserId = userId,
                TotalFamilies = userFamilies.Count(),
                TotalEvents = totalEvents,
                UpcomingEvents = upcomingEvents,
                PastEvents = pastEvents,
                TodayEvents = 0, // Will be calculated separately
                ThisWeekEvents = 0, // Will be calculated separately
                ThisMonthEvents = 0, // Will be calculated separately
                EventsCreatedByUser = eventsCreated,
                EventsUserAttending = eventsAttending,
                ActiveConflicts = 0, // Will be calculated separately
                ResolvedConflicts = 0, // Will be calculated separately
                FamilyStats = new List<FamilyStatisticsSummaryDTO>(),
                EventTypeDistribution = eventTypes,
                DailyActivityPattern = dailyPattern,
                HourlyActivityPattern = hourlyPattern,
                MonthlyActivityPattern = monthlyPattern,
                AverageEventsPerFamily = userFamilies.Any() ? (double)totalEvents / userFamilies.Count() : 0,
                AverageEventsPerWeek = 0, // Will be calculated separately
                EventAttendanceRate = totalEvents > 0 ? (double)eventsAttending / totalEvents * 100 : 0,
                EventCreationRate = totalEvents > 0 ? (double)eventsCreated / totalEvents * 100 : 0,
                LastEventCreated = lastEventCreated,
                LastEventAttended = lastEventAttended,
                NextUpcomingEvent = nextUpcomingEvent,
                BusiestDayOfWeek = dailyPattern.Any() ? (int)Enum.Parse<DayOfWeek>(dailyPattern.OrderByDescending(kvp => kvp.Value).First().Key) : 0,
                BusiestHourOfDay = hourlyPattern.Any() ? hourlyPattern.OrderByDescending(kvp => kvp.Value).First().Key : 9,
                BusiestFamily = "",
                MostCommonEventType = eventTypes.Any() ? eventTypes.OrderByDescending(kvp => kvp.Value).First().Key : "",
                GeneratedAt = DateTime.UtcNow
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting user calendar statistics for user {UserId}", userId);
            throw;
        }
    }

    // Helper methods
    private string GetFamilyColor(int familyId)
    {
        return FamilyColors[familyId % FamilyColors.Length];
    }

    private static bool DoEventsOverlap(FamilyCalendarEvent event1, FamilyCalendarEvent event2)
    {
        return event1.StartTime < event2.EndTime && event2.StartTime < event1.EndTime;
    }

    private static List<TimeSlotDTO> CalculateFreePeriods(DateTime dayStart, DateTime dayEnd, List<TimeSlotDTO> busyPeriods)
    {
        var freePeriods = new List<TimeSlotDTO>();
        var sortedBusy = busyPeriods.OrderBy(b => b.StartTime).ToList();

        var currentTime = dayStart;
        foreach (var busy in sortedBusy)
        {
            if (currentTime < busy.StartTime)
            {
                freePeriods.Add(new TimeSlotDTO
                {
                    StartTime = currentTime,
                    EndTime = busy.StartTime,
                    Description = "Available",
                    Type = "Available"
                });
            }
            currentTime = busy.EndTime > currentTime ? busy.EndTime : currentTime;
        }

        if (currentTime < dayEnd)
        {
            freePeriods.Add(new TimeSlotDTO
            {
                StartTime = currentTime,
                EndTime = dayEnd,
                Description = "Available",
                Type = "Available"
            });
        }

        return freePeriods;
    }

    private static List<TimeSlotDTO> GenerateTimeRecommendations(List<TimeSlotDTO> freePeriods)
    {
        return freePeriods
            .Where(f => (f.EndTime - f.StartTime).TotalMinutes >= 60) // At least 1 hour
            .Take(3)
            .ToList();
    }
} 